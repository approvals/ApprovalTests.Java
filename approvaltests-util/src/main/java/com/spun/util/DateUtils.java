package com.spun.util;

import com.spun.util.logger.SimpleLogger;

import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.TimeZone;

/**
 * A static class of convenience functions for database access
 **/
public class DateUtils
{
  private static int[] TIME_SCALE = {Calendar.YEAR,
                                     Calendar.MONTH,
                                     Calendar.DATE,
                                     Calendar.HOUR_OF_DAY,
                                     Calendar.HOUR,
                                     Calendar.MINUTE,
                                     Calendar.SECOND,
                                     Calendar.MILLISECOND};
  /**
   *
   **/
  public static boolean isSame(Date firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null))
    { return (firstDate == secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime(), smallestUnits);
  }

  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null))
    { return (firstDate == secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime().getTime(), smallestUnits);
  }

  /**
   *
   **/
  public static boolean isSame(Date firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null))
    { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime().getTime(), smallestUnits);
  }

  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null))
    { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime(), smallestUnits);
  }

  /**
   *
   **/
  public static boolean isSame(long firstDate, long secondDate, int smallestUnits)
  {
    if (!ArrayUtils.contains(TIME_SCALE, smallestUnits))
    { throw new Error("Invalid Timescale " + smallestUnits); }
    GregorianCalendar first = new GregorianCalendar();
    first.setTime(new Date(firstDate));
    setSignificantDigit(first, smallestUnits);
    GregorianCalendar second = new GregorianCalendar();
    second.setTime(new Date(secondDate));
    setSignificantDigit(second, smallestUnits);
    //     My_System.variable("Testing if " + first.getTime().getTime() + "==" + second.getTime().getTime());
    return (first.getTime().getTime() == second.getTime().getTime());
  }

  public static Timestamp getStartOfYear()
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.YEAR);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }

  public static Timestamp getStartOfToday()
  {
    return getStartOfXDaysAgo(0);
  }

  public static Timestamp getStartOfXDaysAgo(int numberOfDays)
  {
    return getStartOfXDaysAgo(numberOfDays, new Date());
  }

  public static Timestamp getStartOfXDaysAgo(int numberOfDays, Date startingFrom)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(startingFrom);
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.DATE);
    gregorianCalendar.add(Calendar.DATE, -numberOfDays);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }

  public static Calendar getEndOfTodayAsCalendar()
  {
    return rollToEndOfDay(new Date());
  }

  public static Timestamp getEndOfToday()
  {
    return new Timestamp(getEndOfTodayAsCalendar().getTime().getTime());
  }

  public static Calendar setSignificantDigit(Date date, int smallestUnits)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTimeInMillis(date.getTime());
    return setSignificantDigit(calendar, smallestUnits);
  }

  public static Calendar setSignificantDigit(Calendar calendar, int smallestUnits)
  {
    boolean removeOn = false;
    for (int i = 0; i < TIME_SCALE.length; i++)
    {
      if (removeOn)
      {
        if (TIME_SCALE[i] == Calendar.DATE)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 1);
        }
        else if (TIME_SCALE[i] == Calendar.HOUR_OF_DAY)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 0);
        }
        else
        {
          calendar.clear(TIME_SCALE[i]);
        }
      }
      if (TIME_SCALE[i] == smallestUnits)
      {
        removeOn = true;
      }
    }
    return calendar;
  }

  public static Timestamp getStartOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    return new Timestamp(calendar.getTimeInMillis());
  }

  public static Timestamp getEndOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    calendar.add(unit, 1);
    calendar.add(Calendar.MILLISECOND, -1);
    return new Timestamp(calendar.getTimeInMillis());
  }

  public static boolean areSame(Date date1, Date date2, long accuracy)
  {
    return (compareDates(date1, date2) < accuracy);
  }

  /**
   *
   * {@code return 1 if date1 > date2, 0 if date1 = date2, -1 if date1 < date2 }
   **/
  public static int compareDates(Date date1, Date date2)
  {
    long l1 = toLong(date1);
    long l2 = toLong(date2);
    long diff = l1 - l2;
    return (diff == 0) ? 0 : (int) (diff / Math.abs(diff));
  }

  private static long toLong(Date date)
  {
    return date == null ? 0 : date.getTime();
  }

  public static Date createTime(Date date)
  {
    Calendar time = new GregorianCalendar();
    time.setTime(date);
    time.set(1970, 0, 1);
    return time.getTime();
  }

  /**
   * Rolls back till that time on a 24 hour clock
   **/
  public static Date rollTillHour(int hour, Date date)
  {
    Calendar rolled = new GregorianCalendar();
    rolled.setTime(date);
    setSignificantDigit(rolled, Calendar.HOUR_OF_DAY);
    while (rolled.get(Calendar.HOUR_OF_DAY) != hour)
    {
      rolled.add(Calendar.HOUR_OF_DAY, -1);
    }
    return rolled.getTime();
  }

  public static void main(String[] args)
  {
    SimpleLogger.variable("Calendar.DATE = " + Calendar.DATE);
    SimpleLogger.variable("Year", setSignificantDigit(new GregorianCalendar(), Calendar.YEAR).getTime());
    SimpleLogger.variable("Month", setSignificantDigit(new GregorianCalendar(), Calendar.MONTH).getTime());
    SimpleLogger.variable("Day", setSignificantDigit(new GregorianCalendar(), Calendar.DAY_OF_MONTH).getTime());
    SimpleLogger.variable("Hour", setSignificantDigit(new GregorianCalendar(), Calendar.HOUR).getTime());
    SimpleLogger.variable("Minute", setSignificantDigit(new GregorianCalendar(), Calendar.MINUTE).getTime());
    SimpleLogger.variable("End Of Day", rollToEndOfDay(new Date()).getTime());
  }

  public static GregorianCalendar rollToEndOfDay(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    gregorianCalendar.set(Calendar.HOUR_OF_DAY, 23);
    gregorianCalendar.set(Calendar.MINUTE, 59);
    gregorianCalendar.set(Calendar.SECOND, 59);
    gregorianCalendar.set(Calendar.MILLISECOND, 999);
    return gregorianCalendar;
  }

  public static Timestamp asTimestamp(Date date)
  {
    return new Timestamp(date.getTime());
  }

  public static boolean isToday(Date date)
  {
    return DateUtils.isSame(date, new Date(), Calendar.DATE);
  }

  public static Calendar asCalendar(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    return gregorianCalendar;
  }

  public static Timestamp getLastOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, false);
  }

  public static Timestamp getNextOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, true);
  }

  private static Timestamp getDayOfWeek(int dayOfWeek, boolean foward)
  {
    int multiplier = foward ? -1 : 1;
    for (int i = 0; i < 7; i++)
    {
      Timestamp day = getStartOfXDaysAgo(i * multiplier);
      if (asCalendar(day).get(Calendar.DAY_OF_WEEK) == dayOfWeek)
      { return day; }
    }
    throw new Error(String.format("didn't find a %s in the %s 7 days", dayOfWeek, foward ? "next" : "last"));
  }

  /**
   * @param date "yyyy/MM/dd"
   */
  public static Timestamp parse(String date)
  {
    SimpleDateFormat format = new java.text.SimpleDateFormat("yyyy/MM/dd");
    try
    {
      return asTimestamp(format.parse(date));
    }
    catch (ParseException e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }

  public static boolean doesDaylightSavingsTimeStartOn(String date)
  {
    Timestamp day = parse(date);
    TimeZone timeZone = GregorianCalendar.getInstance().getTimeZone();
    boolean inDaylightTime = timeZone.inDaylightTime(day);
    Calendar tomorrow = asCalendar(day);
    tomorrow.add(Calendar.DATE, 1);
    Date time = tomorrow.getTime();
    return !inDaylightTime && timeZone.inDaylightTime(time);
  }

  public static Date toDateInLocalTimeZone(LocalDateTime dateTime)
  {
    return toDate(dateTime, ZoneId.systemDefault());
  }

  public static Date toDateInUTC(LocalDateTime dateTime)
  {
    return toDate(dateTime, ZoneOffset.UTC);
  }

  public static Date toDate(LocalDateTime dateTime, ZoneId zoneId)
  {
    return Date.from(dateTime.atZone(zoneId).toInstant());
  }

  public static Calendar setTime(Calendar day, int hour, int minute)
  {
    return new GregorianCalendar(day.get(Calendar.YEAR), day.get(Calendar.MONTH), day.get(Calendar.DAY_OF_MONTH),
        hour, minute);
  }

  public static boolean isEndOfTime(Date time)
  {
    return time.getTime() == Long.MAX_VALUE;
  }

  public static String toString(Date date)
  {
    if (date == null)
    {
      return "null";
    }
    else if (isEndOfTime(date))
    {
      return "<EndOfTime>";
    }
    else if (date.getTime() == 0)
    {
      return "<Epoch>";
    }
    else
    {
      return asTimestamp(date).toString();
    }
  }
}
