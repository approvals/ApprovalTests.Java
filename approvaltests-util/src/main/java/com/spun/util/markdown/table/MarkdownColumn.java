package com.spun.util.markdown.table;

public enum MarkdownColumn {
                            DEFAULT, RIGHT_JUSTIFIED, LEFT_JUSTIFIED
}
