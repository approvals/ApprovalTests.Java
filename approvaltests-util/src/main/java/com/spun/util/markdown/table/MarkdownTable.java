package com.spun.util.markdown.table;

import com.spun.util.markdown.MarkdownCompatible;
import org.lambda.functions.Function1;
import org.lambda.query.Queryable;

public class MarkdownTable implements MarkdownCompatible
{
  public Queryable<MarkdownTableElement> markdown         = new Queryable<MarkdownTableElement>(
      MarkdownTableElement.class);
  private boolean                        fixedWidth       = true;
  private MarkdownColumn                 columnProperties = MarkdownColumn.DEFAULT;
  public void setColumnsConsistentWidth(boolean setting)
  {
    fixedWidth = setting;
  }

  public static <I, O> MarkdownTable create(I[] inputs, Function1<I, O> o, String column1, String column2)
  {
    MarkdownTable table = new MarkdownTable().withColumnHeaders(column1, column2);
    for (I input : inputs)
    {
      table.addRow(input, o.call(input));
    }
    return table;
  }

  public static MarkdownTable withHeaders(String... columnNames)
  {
    MarkdownTable table = new MarkdownTable();
    return table.withColumnHeaders(columnNames);
  }

  public <I> MarkdownTable addRowsForInputs(I[] inputs, Function1<I, Object>... transfers)
  {
    for (I input : inputs)
    {
      Queryable<Object> row = Queryable.as(transfers).select(f -> f.call(input));
      row.add(0, input);
      addRow(row.toArray());
    }
    return this;
  }

  public MarkdownTable addRow(Object... columns)
  {
    markdown.addAll(constructRow(columns));
    return this;
  }

  public MarkdownTable withColumnHeaders(String... headers)
  {
    markdown.addAll(constructColumnHeaders(headers));
    return this;
  }

  @Override
  public String toString()
  {
    return toMarkdown();
  }

  @Override
  public String toMarkdown()
  {
    handleFixedWidth();
    return render(markdown);
  }

  private void handleFixedWidth()
  {
    if (!fixedWidth)
    { return; }
    Queryable<Queryable<MarkdownTableElement>> rows = markdown.split(e -> e == MarkdownTableElement.NEWLINE);
    for (int column = 0; column < rows.get(0).size(); column++)
    {
      int col = column;
      if (rows.get(0).get(column) instanceof MarkdownTableContents)
      {
        Queryable<Resizable> select = rows.select(e -> e.get(col)).where(e -> e instanceof Resizable)
            .select(e -> (Resizable) e);
        select.forEach(e -> e.setJustification(columnProperties));
        int length = select.max(e -> e.getLength()).getLength();
        select.forEach(e -> e.setPadding(length));
      }
    }
  }

  public static String printColumnHeaders(String... headers)
  {
    return render(constructColumnHeaders(headers));
  }

  public static Queryable<MarkdownTableElement> constructColumnHeaders(String... headers)
  {
    Queryable<MarkdownTableElement> row = constructRow(headers);
    for (int i = 0; i < headers.length; i++)
    {
      row.add(MarkdownTableElement.DELIMITER);
      row.add(new MarkdownTableHeader());
    }
    row.add(MarkdownTableElement.DELIMITER);
    row.add(MarkdownTableElement.NEWLINE);
    return row;
  }

  public static String printRow(Object... columns)
  {
    return render(constructRow(columns));
  }

  private static String render(Queryable<MarkdownTableElement> table)
  {
    return table.join("");
  }

  public static Queryable<MarkdownTableElement> constructRow(Object... columns)
  {
    Queryable row = new Queryable(MarkdownTableElement.class);
    row.add(MarkdownTableElement.DELIMITER);
    for (int x = 0; x < columns.length; ++x)
    {
      row.add(new MarkdownTableContents("" + columns[x]));
      row.add(MarkdownTableElement.DELIMITER);
    }
    row.add(MarkdownTableElement.NEWLINE);
    return row;
  }

  public void setColumnProperties(MarkdownColumn columnProperties)
  {
    this.columnProperties = columnProperties;
  }
}
