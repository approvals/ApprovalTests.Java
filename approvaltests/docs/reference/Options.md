<a id="top"></a>

# Options

<!-- toc -->
## Contents

  * [Introduction](#introduction)
  * [Fluent Interface](#fluent-interface)
  * [Reporters](#reporters)
  * [Scrubbers](#scrubbers)
  * [File Options](#file-options)
    * [File Extensions](#file-extensions)
  * [Custom Comparators](#custom-comparators)
  * [Defaults](#defaults)
  * [Adding to Existing Options object](#adding-to-existing-options-object)<!-- endToc -->

## Introduction

There are many things you might want to tweak with Approval Tests. `Options` is the entry-point for many of the changes.
It is on all `verify()` methods, as an optional parameter.

## Fluent Interface

`Options` utilizes a fluent interface, allowing you to chain together commands. Each returned object is a new copy.

<!-- snippet: specify_all_the_options -->
<a id='snippet-specify_all_the_options'></a>
```java
new Options().withReporter(new ReportNothing()).withScrubber(new GuidScrubber()).forFile()
    .withExtension(".json");
```
<sup><a href='/approvaltests-tests/src/test/java/org/approvaltests/OptionsSamplesTest.java#L15-L18' title='Snippet source file'>snippet source</a> | <a href='#snippet-specify_all_the_options' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## Reporters

[Reporters](/docs/reference/Reporters.md#top) launch diff tools upon failure.  Read how to configure them with `Options` [here](Reporters.md#via-options):

## Scrubbers

[Scrubbers](/docs/explanations/Scrubbers.md#top) clean output to help remove inconsistent pieces of text, such as dates.  Read how to set up Scrubbers [here](../Scrubbers.md#configuring-scrubbers):
There are two ways to set a Scrubber.

## File Options

The `Options.forFile()` class exists to customise the `.approved` and `.received` files in various ways.

### File Extensions

If you want to change the file extension of both the approved and received files, use `withExtension()`.

<!-- snippet: basic_approval_with_file_extension -->
<a id='snippet-basic_approval_with_file_extension'></a>
```java
Approvals.verify("text to be verified", new Options().forFile().withExtension(".xyz"));
```
<sup><a href='/approvaltests-tests/src/test/java/org/approvaltests/OptionsSamplesTest.java#L23-L25' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic_approval_with_file_extension' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

**Note:** `withExtension()` returns an `Options` object, so it's possible to keep appending more `with...()` calls.

## Custom Comparators
By default Approval Tests will only check if the `.approved` and `.received` files are _exactly_ matching.
The only accomodations it makes is for differences in line endings.

If you would like to create a more flexible comparison, you can do it by using the `Options.withComparator()` function.

## Defaults

The default constructor for Options does:

* no scrubbing
* uses file extension `.txt`
* uses whatever is currently set as the default reporter.

## Adding to Existing Options object

Each instance of Options is not mutable. Therefore, every call produces a new copy of the Options object. This allows you to modify a specific option from an existing Options object,
while retaining all other settings, and not changing the original object.


---

[Back to User Guide](../README.md#top)
