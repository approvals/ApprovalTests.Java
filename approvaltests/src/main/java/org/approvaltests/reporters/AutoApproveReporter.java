package org.approvaltests.reporters;

import com.spun.util.io.FileUtils;
import org.approvaltests.core.VerifyResult;

import java.io.File;

public class AutoApproveReporter implements ReporterWithApprovalPower
{
  @Override
  public boolean report(String received, String approved)
  {
    File a = new File(approved);
    if (a.exists())
    {
      a.delete();
    }
    File r = new File(received);
    FileUtils.copyFile(r, a);
    return true;
  }

  @Override
  public VerifyResult approveWhenReported()
  {
    return VerifyResult.SUCCESS;
  }
}
