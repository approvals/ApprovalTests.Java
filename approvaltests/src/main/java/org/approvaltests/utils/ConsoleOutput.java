package org.approvaltests.utils;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class ConsoleOutput implements AutoCloseable
{
  private final ByteArrayOutputStream outContent;
  private final ByteArrayOutputStream errContent;
  private final PrintStream           originalOut;
  private final PrintStream           originalErr;

  public ConsoleOutput()
  {
    originalOut = System.out;
    originalErr = System.err;
    outContent = new ByteArrayOutputStream();
    errContent = new ByteArrayOutputStream();
    System.setOut(new PrintStream(outContent));
    System.setErr(new PrintStream(errContent));
  }

  public String getOutput()
  {
    return outContent.toString();
  }

  public String getError()
  {
    return errContent.toString();
  }

  public void verifyOutput()
  {
    verifyOutput(new Options());
  }

  public void verifyOutput(Options options)
  {
    Approvals.verify(getOutput(), options);
  }

  public void verifyError()
  {
    Approvals.verify(getError());
  }

  public void verifyError(Options options)
  {
    Approvals.verify(getError(), options);
  }

  public void verifyAll()
  {
    verifyAll(new Options());
  }

  public void verifyAll(Options options)
  {
    String combined = "Output:\n" + getOutput() + "\nError:\n" + getError();
    Approvals.verify(combined, options);
  }

  @Override
  public void close()
  {
    System.setOut(originalOut);
    System.setErr(originalErr);
  }
}