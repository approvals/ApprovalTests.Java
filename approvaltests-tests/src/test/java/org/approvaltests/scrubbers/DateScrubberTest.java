package org.approvaltests.scrubbers;

import com.spun.util.DateUtils;
import com.spun.util.markdown.table.MarkdownColumn;
import com.spun.util.markdown.table.MarkdownTable;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.utils.ConsoleOutput;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.reporters.AutoApproveReporter;

public class DateScrubberTest
{
  @Test
  void testSupportedFormatWorksForExamples()
  {
    for (DateScrubber.SupportedFormat supportedFormat : DateScrubber.getSupportedFormats())
    {
      DateScrubber dateScrubber = new DateScrubber(supportedFormat.getRegex());
      for (String example : supportedFormat.getExamples())
      {
        assertEquals("[Date1]", dateScrubber.scrub(example),
            () -> "didn't work for regex " + supportedFormat.getRegex());
      }
    }
  }
  @Test
  void testGetDateScrubber()
  {
    List<String> formats = Stream.of(DateScrubber.getSupportedFormats()).flatMap(f -> Stream.of(f.getExamples()))
        .collect(Collectors.toList());
    Approvals.verifyAll("Date scrubbing", formats, this::verifyScrubbing);
  }
  private String verifyScrubbing(String formattedExample)
  {
    DateScrubber scrubber = DateScrubber.getScrubberFor(formattedExample);
    String exampleText = String.format("{'date':\"%s\"}", formattedExample);
    return String.format("Scrubbing for %s:\n" + "%s\n" + "Example: %s\n\n", formattedExample, scrubber,
        scrubber.scrub(exampleText));
  }
  @Test
  void exampleForDocumentation()
  {
    // begin-snippet: scrub-date-example
    Approvals.verify("created at 03:14:15", new Options().withScrubber(DateScrubber.getScrubberFor("00:00:00")));
    // end-snippet
  }
  @Test
  void supportedFormats()
  {
    String[] headers = {"Example Date", "RegEx Pattern"};
    MarkdownTable table = MarkdownTable.withHeaders(headers);
    table.setColumnProperties(MarkdownColumn.LEFT_JUSTIFIED);
    for (DateScrubber.SupportedFormat format : DateScrubber.getSupportedFormats())
    {
      table.addRow(format.getExamples()[0], format.getRegex());
    }
    Approvals.verify(String.format("\n\n%s\n\n", table.toMarkdown()),
        new Options().forFile().withExtension(".md"));
  }
  @Disabled("use when new examples are shared at https://github.com/approvals/ApprovalTests.Java/issues/112")
  @Test
  void textExamples()
  {
    String[] examples = """
        2025-05-15 16:57:04.599
        """.split("\n");
    Approvals.verifyAll("Date scrubbing", examples, this::verifyScrubbing);
  }
  @Test
  void testBlank()
  {
    DateScrubber.getScrubberFor("");
  }
  @Test
  void testTimestamp()
  {
    String input = DateUtils.asTimestamp(new Date()).toString();
    String scrubbed = DateScrubber.getScrubberForTimestamp().scrub(input);
    assertEquals("[Date1]", scrubbed);
  }
  @Test
  void testSqlDate()
  {
    java.sql.Date sqlDate = new java.sql.Date(System.currentTimeMillis());
    String input = sqlDate.toString();
    String scrubbed = DateScrubber.getScrubberForSqlDate().scrub(input);
    assertEquals("[Date1]", scrubbed);
  }
  @Test
  void testUtilDate()
  {
    java.util.Date utilDate = new java.util.Date();
    String input = utilDate.toString();
    String scrubbed = DateScrubber.getScrubberForDate().scrub(input);
    assertEquals("[Date1]", scrubbed);
  }
  @Test
  void testAddScrubberWithValidRegexAndExample()
  {
    try
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}");
      DateScrubber scrubber = DateScrubber.getScrubberFor("2024-Jan-15");
      assertEquals("[Date1]", scrubber.scrub("2024-Jan-15"));
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
  @Test
  void testAddScrubberWithInvalidRegex()
  {
    assertThrows(IllegalArgumentException.class, () -> {
      DateScrubber.addScrubber("2023-Dec-25", "[invalid regex");
    });
  }
  @Test
  void testAddScrubberWithRegexThatDoesntMatchExample()
  {
    assertThrows(IllegalArgumentException.class, () -> {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{2}-\\d{2}-\\d{4}");
    });
  }
  @Test
  @UseReporter(AutoApproveReporter.class)
  void testAddScrubberDisplaysMessage()
  {
    var expected = """
        You are using a custom date scrubber.
        If you think the format you want to scrub would be useful for others,
        please add it to
        https://github.com/approvals/ApprovalTests.Java/issues/112.

        To suppress this message, use:
            DateScrubber.addScrubber("<date format>", "<regex>", false)
        """;
    try (ConsoleOutput console = new ConsoleOutput())
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}");
      console.verifyOutput(new Options().inline(expected));
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
  @Test
  void testAddScrubberSuppressMessage()
  {
    try (ConsoleOutput console = new ConsoleOutput())
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}", false);
      String output = console.getOutput();
      assertEquals("", output);
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
  @Test
  void testCustomScrubberIntegrationWithGetScrubberFor()
  {
    try
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}", false);
      DateScrubber scrubber = DateScrubber.getScrubberFor("2024-Jan-15");
      assertEquals("[Date1]", scrubber.scrub("2024-Jan-15"));
      assertEquals("[Date1]", scrubber.scrub("2025-Mar-30"));
      assertEquals("Meeting on [Date1] at noon", scrubber.scrub("Meeting on 2024-Jan-15 at noon"));
      assertEquals("Due date: [Date1]", scrubber.scrub("Due date: 2025-Mar-30"));
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
  @Test
  void testClearCustomScrubbers()
  {
    try
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}", false);
      DateScrubber scrubber = DateScrubber.getScrubberFor("2024-Jan-15");
      assertEquals("[Date1]", scrubber.scrub("2024-Jan-15"));
      DateScrubber.clearCustomScrubbers();
      assertThrows(Exception.class, () -> {
        DateScrubber.getScrubberFor("2024-Jan-15");
      });
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
  @Test
  void testMultipleCustomScrubbers()
  {
    try
    {
      DateScrubber.addScrubber("2023-Dec-25", "\\d{4}-[A-Za-z]{3}-\\d{2}", false);
      DateScrubber.addScrubber("25/Dec/2023", "\\d{2}/[A-Za-z]{3}/\\d{4}", false);
      DateScrubber scrubber1 = DateScrubber.getScrubberFor("2024-Jan-15");
      assertEquals("[Date1]", scrubber1.scrub("2024-Jan-15"));
      DateScrubber scrubber2 = DateScrubber.getScrubberFor("15/Jan/2024");
      assertEquals("[Date1]", scrubber2.scrub("15/Jan/2024"));
    }
    finally
    {
      DateScrubber.clearCustomScrubbers();
    }
  }
}