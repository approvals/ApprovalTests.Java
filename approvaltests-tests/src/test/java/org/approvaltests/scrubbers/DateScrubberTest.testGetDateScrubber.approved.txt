Date scrubbing


Scrubbing for Tue May 13 16:30:00:
RegExScrubber[[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2}]
Example: {'date':"[Date1]"}


Scrubbing for Wed Nov 17 22:28:33 EET 2021:
RegExScrubber[[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2} [a-zA-Z]{3,4} \d{4}]
Example: {'date':"[Date1]"}


Scrubbing for Wed, 21 Oct 2015 07:28:00 GMT:
RegExScrubber[(Mon|Tue|Wed|Thu|Fri|Sat|Sun), \d{2} (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d{4} \d{2}:\d{2}:\d{2} GMT]
Example: {'date':"[Date1]"}


Scrubbing for Tue May 13 2014 23:30:00.789:
RegExScrubber[[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{4} \d{2}:\d{2}:\d{2}.\d{3}]
Example: {'date':"[Date1]"}


Scrubbing for Tue May 13 16:30:00 -0800 2014:
RegExScrubber[[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2} -\d{4} \d{4}]
Example: {'date':"[Date1]"}


Scrubbing for 13 May 2014 23:50:49,999:
RegExScrubber[\d{2} [a-zA-Z]{3} \d{4} \d{2}:\d{2}:\d{2},\d{3}]
Example: {'date':"[Date1]"}


Scrubbing for Oct 13 15:29:
RegExScrubber[[A-Za-z]{3} \d{2} \d{2}:\d{2}]
Example: {'date':"[Date1]"}


Scrubbing for May 13, 2014 11:30:00 PM PST:
RegExScrubber[[a-zA-Z]{3} \d{2}, \d{4} \d{2}:\d{2}:\d{2} [a-zA-Z]{2} [a-zA-Z]{3}]
Example: {'date':"[Date1]"}


Scrubbing for 23:30:00:
RegExScrubber[\d{2}:\d{2}:\d{2}]
Example: {'date':"[Date1]"}


Scrubbing for 2014/05/13 16:30:59.786:
RegExScrubber[\d{4}/\d{2}/\d{2} \d{2}:\d{2}:\d{2}.\d{2}\d]
Example: {'date':"[Date1]"}


Scrubbing for 2020-9-10T08:07Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2020-09-9T08:07Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2020-09-10T8:07Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2020-09-10T08:07Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2020-09-10T08:07:89Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}:\d{2}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2020-09-10T01:23:45.678Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}\:\d{2}\.\d{3}Z]
Example: {'date':"[Date1]"}


Scrubbing for 20210505T091112Z:
RegExScrubber[\d{8}T\d{6}Z]
Example: {'date':"[Date1]"}


Scrubbing for 2024-12-17:
RegExScrubber[\d{4}-\d{2}-\d{2}]
Example: {'date':"[Date1]"}


Scrubbing for 2024-12-18T14:04:46.746130Z:
RegExScrubber[\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}:\d{2}.\d{1,6}Z]
Example: {'date':"[Date1]"}


